scipy/_lib/_uarray/_uarray_dispatch.cxx

```diff
@@ -1775,13 +1775,8 @@ PyModuleDef uarray_module = {
 
 } // namespace
 
-#if defined(WIN32) || defined(_WIN32)
-#  define MODULE_EXPORT __declspec(dllexport)
-#else
-#  define MODULE_EXPORT __attribute__((visibility("default")))
-#endif
-
-extern "C" MODULE_EXPORT PyObject * PyInit__uarray(void) {
+PyMODINIT_FUNC
+PyInit__uarray(void) {
```

Comment:
This is not working on the `Python3.8-dbg` Linux job:
```
ImportError: dynamic module does not define module export function (PyInit__uarray)
```
That's for tomorrow ....

Reply:
This is still running on distutils, not Meson, right?

Cross-posting from the Meson PR:

Interesting to note: exactly one of the 19 files you fixed, manually coded this (reimplementing PyMODINIT_FUNC but without cygwin support and without support for unix compilers other than gcc/clang):
```c

#if defined(WIN32) || defined(_WIN32)
#  define MODULE_EXPORT __declspec(dllexport)
#else
#  define MODULE_EXPORT __attribute__((visibility("default")))
#endif

extern "C" MODULE_EXPORT PyObject * PyInit__uarray(void) {
```

So, something somewhere is trying to compile `_uarray_dispatch.cxx` with `-fvisibility=hidden`, but, on Python3.8 the PyMODINIT_FUNC does not yet mark the init symbol as default (public) visibility.

This is why after testing, I had to make Meson only set `-fvisibility=hidden` for Python3.9 and on.

Reply:
Three modules are particularly affected by this:
- the broken one
  https://github.com/scipy/scipy/blob/8a64c938ddf1ae4c02a08d2c5e38daeb8d061d38/scipy/_lib/_uarray/setup.py#L13

- the pybind11 one
  https://github.com/scipy/scipy/blob/8a64c938ddf1ae4c02a08d2c5e38daeb8d061d38/scipy/fft/_pocketfft/setup.py#L26
  https://github.com/scipy/scipy/blob/8a64c938ddf1ae4c02a08d2c5e38daeb8d061d38/scipy/fft/_pocketfft/pypocketfft.cxx#L716

- the other pybind11 one
  https://github.com/scipy/scipy/blob/ecb800f37bc6c6eb4f160cba34fec81932eaea70/scipy/spatial/setup.py#L18
  https://github.com/scipy/scipy/blob/ecb800f37bc6c6eb4f160cba34fec81932eaea70/scipy/spatial/src/distance_pybind.cpp#L550


Note that pybind11 sets up visibility without regard to the Python.h header, so it works correctly regardless of python version, therefore it's perfectly fine for distutils to pass `-fvisibility=hidden` there.

For _uarray, the manually coded visibility attributes *also* worked regardless of Python version, which is why the passing `-fvisibility=hidden` there produced correct results. Moving to `PyMODINIT_FUNC` means that _uarray, specifically, is no longer performing symbol force-exporting on python less than 3.9, hence it is no longer safe to do symbol hiding in distutils... and therefore the breakage.

You can use multiple attributes at the same time, so perhaps it would make sense to partially roll back this specific diff hunk and just use

```c
MODULE_EXPORT PyMODINIT_FUNC
PyInit__uarray(void) { 
```

Reply:
Thanks for all the details @eli-schwartz!

> `MODULE_EXPORT PyMODINIT_FUNC`

That didn't quite work, that gives:
```
/home/rgommers/anaconda3/envs/scipy-dev-py38/include/python3.8/pyport.h:690:46: error: expected unqualified-id before string constant
  690 | #               define PyMODINIT_FUNC extern "C" PyObject*
      |                                              ^~~
scipy/_lib/_uarray/_uarray_dispatch.cxx:1786:15: note: in expansion of macro 'PyMODINIT_FUNC'
 1786 | MODULE_EXPORT PyMODINIT_FUNC
      |               ^~~~~~~~~~~~~~
```
I couldn't get a single invocation to work, so went with a simple switch on Python version, keeping the Python 3.8 flavor unchanged.

Reply:
Ah right... This has some CXX goop in the middle (extern "C") so the attributes aren't a single attribute list and that won't work. It would work for C, because that is just `<attribute> PyObject*` and you can cleanly prepend to that. But basically it's conditional on the exact text of the macro expansion, so maybe not really worth it...